File: ruby/proverb/proverb.rb
class Proverb
  def initialize(*strings)
    @first = strings.first
    @strings = strings
  end

  def to_s
    proverb = []
    @strings.each_with_index do |str, i|
      proverb << "For want of a #{str} the #{@strings[i + 1]} was lost."
    end
    proverb.pop
    proverb.push("And all for the want of a #{@first}.").join("\n")
  end
end


# Proverb.new('nail', 'shoe').to_s
# p Proverb.new('nail', 'shoe', 'horse').to_s

File: javascript/proverb/proverb.js
const line = (first, second) => {
  return typeof second === 'undefined' ? '' : `For want of a ${first} the ${second} was lost.\n`;
};

const proverb = (...input) => {
  const proverbArray = [];

  input.forEach((value, index) => {
    proverbArray.push(line(value, input[index + 1]));
  });

  const last = input[input.length - 1];
  let qualifier = '';

  if (last.qualifier) {
    qualifier = `${last.qualifier} `;
    proverbArray.length -= 2;
  }

  proverbArray.push(`And all for the want of a ${qualifier}${input[0]}.`);

  return proverbArray.join('');
};

if (module) { module.exports = proverb; }
